lib/ui/painting.dart

```diff
@@ -4935,26 +4591,1632 @@ class PictureRecorder extends NativeFieldWrapperClass2 {
+      } else {
+        _addOp(_CanvasOp.setImageFilter, 0);
+        _addImageFilter(filter);
+      }
+      _curImageFilter = paint.imageFilter;
+    }
+    if ((dataNeeded & _maskFilterNeeded) != 0 && _curMaskFilter != paint.maskFilter) {
+      final MaskFilter? filter = paint.maskFilter;
+      if (filter == null) {
+        _addOp(_CanvasOp.clearMaskFilter, 0);
+      } else {
+        _addOp(_maskFilterOps[filter._style.index], 1);
+        _addScalar(filter._sigma);
+      }
+      _curMaskFilter = paint.maskFilter;
+    }
+  }
+
+  @override
+  void drawPaint(Paint paint) {
+    _updatePaintData(paint, _paintMask);
+    _addOp(_CanvasOp.drawPaint, 0);
+  }
+
+  @override
+  void drawColor(Color color, BlendMode blendMode) {
+    _addOp(_CanvasOp.drawColor, 2);
+    _addInt(color.value);
+    _addInt(blendMode.index);
+  }
+
+  @override
+  void drawLine(Offset p1, Offset p2, Paint paint) {
+    _updatePaintData(paint, _strokeMask);
+    _addOp(_CanvasOp.drawLine, 2 * _nOffsetData);
+    _addOffset(p1);
+    _addOffset(p2);
+    _addBounds(Rect.fromPoints(p1, p2), true);
+  }
+
+  @override
+  void drawRect(Rect rect, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(_CanvasOp.drawRect, _nRectData);
+    _addRect(rect);
+    _addBounds(rect);
+  }
+
+  @override
+  void drawOval(Rect rect, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(_CanvasOp.drawOval, _nRectData);
+    _addRect(rect);
+    _addBounds(rect);
+  }
+
+  @override
+  void drawCircle(Offset center, double radius, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(_CanvasOp.drawCircle, _nOffsetData + 1);
+    _addOffset(center);
+    _addScalar(radius);
+    _addBounds(Rect.fromCenter(center: center, width: radius, height: radius));
+  }
+
+  @override
+  void drawRRect(RRect rrect, Paint paint) {
+    final Rect outerRect = rrect.outerRect;
+    if (rrect.isRect) {
+      drawRect(outerRect, paint);
+    } else if (rrect.isEllipse) {
+      drawOval(outerRect, paint);
+    } else {
+      _updatePaintData(paint, _drawMask);
+      _addOp(_CanvasOp.drawRRect, _nRoundRectData);
+      _addRRect(rrect);
+    }
+    _addBounds(outerRect);
+  }
+
+  @override
+  void drawDRRect(RRect outer, RRect inner, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(_CanvasOp.drawDRRect, 2 * _nRoundRectData);
+    _addRRect(outer);
+    _addRRect(inner);
+    _addBounds(outer.outerRect);
+  }
+
+  @override
+  void drawArc(Rect rect, double startAngle, double sweepAngle, bool useCenter, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(useCenter ? _CanvasOp.drawArcCenter : _CanvasOp.drawArc, _nRectData + 2);
+    _addRect(rect);
+    _addScalar2(startAngle, sweepAngle);
+    _addBounds(rect);
+  }
+
+  @override
+  void drawPath(Path path, Paint paint) {
+    _updatePaintData(paint, _drawMask);
+    _addOp(_CanvasOp.drawPath, 0);
+    _addPathData(path);
+    _addBounds(path.getBounds());
```

Comment:
We may want to optimize `path.getBounds` if we're going to calling it so often. It's not a cheap method.

We also don't guarantee that path.getBounds is correct. See https://api.flutter.dev/flutter/dart-ui/Path/getBounds.html

Reply:
`Path.getBounds` overestimates its size, which is fine for the purposes of computing the bounds of the picture (I'm assuming this is to be used for optimization, such as pruning of paint ops). This is also the case on the web. Example:

https://github.com/flutter/goldens/blob/master/engine/web/paint_bounds_for_line_rotated.png

Reply:
Yeah, I think the case where it gives you a rect that's 0 on the x or y dimension is probably more concerning.

Reply:
Does it do that if you give it a moveTo()? Isn't the example you gave above correct based on an assumed moveTo(0,0)?

Reply:
In looking back on this, the bounds aren't really needed in Dart anyway (though I think we should provide them lazily on request). I think all of this accumulation was actually for when we add it as a layer, but the layer code could do a native C++ run through of the data and compute the bounds itself. That would get rid of both the ctm and the accumulator from the Dart code.

Reply:
Paths have an implicit `moveTo(0, 0)`. So the bounds come out the same with or without `moveTo(0, 0)`. If you move to some other point, that affects the bounds.

Reply:
I still don't understand your point about the bounds. The bounds you gave above were correct. You need to include a moveTo in the bounds (unless there are multiple back to back moveTo calls in which case the Path object should be smart enough to ignore the duplicates. What are you concerned about here?

(Also, this code may move to the native code, but the same issue would likely happen there since we are using the same backing object to compute the bounds whether we call it from Dart or C++)

Reply:
I'm concerned that a picture that's meant to represent a hairline would get optimized away because it appear to have no width (or no height depending on the dimension).

Moving this to native would have the advantage of not needing multiple hops back and forth to native to get this information.

Reply:
A path cannot represent a hairline. You don't know the width until you render it as that information is in the Paint object.

With a strokeWidth of 2000, that line from 0,0 to 10,10 would fill the screen. The Path has no idea.

Bounds of paths are a rat's nest and there are lots of tradeoffs between speed and accuracy.